presto-main/src/main/java/com/facebook/presto/metadata/PolymorphicScalarFunctionBuilder.java

```diff
@@ -220,56 +195,140 @@ public MethodsGroupBuilder methods(List<String> methodNames)
             requireNonNull(methodNames, "methodNames is null");
             checkArgument(!methodNames.isEmpty(), "methods list is empty");
 
-            List<Method> matchingMethods = asList(clazz.getMethods()).stream()
+            List<MethodAndNativeContainerTypes> matchingMethods = asList(clazz.getMethods()).stream()
                     .filter(method -> methodNames.contains(method.getName()))
+                    .map(method -> new MethodAndNativeContainerTypes(method, nCopies(signature.getArgumentTypes().size(), Optional.empty())))
                     .collect(toImmutableList());
             List<String> matchingMethodNames = matchingMethods.stream()
-                    .map(Method::getName)
+                    .map(methodAndNativeContainerTypes -> methodAndNativeContainerTypes.getMethod().getName())
                     .collect(toImmutableList());
 
             for (String methodName : methodNames) {
                 checkState(matchingMethodNames.contains(methodName), "method %s was not found in %s", methodName, clazz);
             }
 
-            this.methods = matchingMethods;
+            this.methodAndNativeContainerTypes = matchingMethods;
```

Comment:
I propose:

* Change this to `this.methodAndNativeContainerTypes.addAll(matchingMethods)`
* Make `methodWithExplicitJavaTypes` add new methods (instead of modifying existing ones).
* When the above is done, it would make sense to make the two `methods` methods delegate to `methodWithExplicitJavaTypes` method.

In general, it seems very strange that `setExplicitNativeContainerTypes` exists on `MethodAndNativeContainerTypes`